### PR TITLE
Change the return type of CalculateNumChildren to uint32_t.

### PR DESCRIPTION
**Pull request by adrian-prantl**
In the end this value comes from TypeSystem::GetNumChildren which returns a uint32_t, so ValueObject should be consistent with that.

### CODE REVIEW

**Review by kastiglione — APPROVED** [review]

**Comment by jimingham** [comment]
Theoretically someone could make a synthetic type or an array with more than UINT32_MAX elements, and with this change we can't support displaying that.  I did get one bug report that this didn't work a few years back, though this likely a very uncommon case.  If it were easy to go the other way, and pipe num_children as a size_t meaningfully through clang and lldb, that would be formally better.
But that's a nice to have, and if the clang changes would be disruptive, then this change is also fine by me.  Better to not lie to ourselves about what we can do.

**Comment by bulbazord** [comment]
Why not increase TypeSystem::GetNumChildren to return a size_t instead?

**Comment by adrian-prantl** [comment]
> Why not increase TypeSystem::GetNumChildren to return a size_t instead?

We could, but even 2^32 seems like an outrageous amount of children to me. What would be a use-case for this? A ValueObject that has the entire address space as synthetic children?

There is other code that assumes that we can store the result of GetNumChildren() in an int64 and use negative numbers as out-of-band signaling. We could replace that with std::optional of course.

**Comment by jimingham** [comment]
I got one bug report - which I can't currently find - from someone who had managed to make a thing this big and we wouldn't print elements out of it correctly.  We didn't do anything about it at the time, it did seem silly.

But it WAS actually silly, at least once.

Jim

**Comment by JDevlieghere** [comment]
> > Why not increase TypeSystem::GetNumChildren to return a size_t instead?
> 
> We could, but even 2^32 seems like an outrageous amount of children to me. What would be a use-case for this? A ValueObject that has the entire address space as synthetic children?
> 
> There is other code that assumes that we can store the result of GetNumChildren() in an int64 and use negative numbers as out-of-band signaling. We could replace that with std::optional of course.

But what's there to gain by limiting it to 32 bits instead of 64? Saying it's unlikely or overkill isn't a very strong argument. I think a possibly better answer to Alex' question is probably: because `SBValue::GetNumChildren` returns a `uint32_t` and we cannot change that, so we'd have to check for overflow at SB API level. I don't feel strongly either way (though if it were me, I'd make everything a `size_t`.)

**Review by clayborg — COMMENTED** [review]
I would almost vote to change everything to `uint64_t` except for the public API since we can't change the API without breaking. Though I winder if we can actually change this one:
```
uint64_t SBValue::GetNumChildren();
```
Since the return value isn't mangled into the function name?

The reason I mention the uint64_t is `lldb::SBValue` and `lldb_private::ValueObject*` can represent _any_ object that can be expanded. We could have a `lldb::SBValue` that represents all of memory in a process where each object can represent an area in memory as a specific format and size. So a 64 bit process _could_ have a `SBValue` with `UINT64_MAX` children available if we wanted to have a `SBValue` that represented memory as `uint8_t` objects. 

So if we are going to change stuff around, I would vote to use `uint64_t` instead of `uint32_t`

**Comment by clayborg** [comment]
I would vote for `uint64_t` over `size_t` as `size_t` is 32 bits on 32 bit operating systems and we might be cross debugging to 64 bit systems that need `uint64_t`

**Comment by bulbazord** [comment]
> I would almost vote to change everything to `uint64_t` except for the public API since we can't change the API without breaking. Though I winder if we can actually change this one:
> 
> ```
> uint64_t SBValue::GetNumChildren();
> ```
> 
> Since the return value isn't mangled into the function name?
> 
> The reason I mention the uint64_t is `lldb::SBValue` and `lldb_private::ValueObject*` can represent _any_ object that can be expanded. We could have a `lldb::SBValue` that represents all of memory in a process where each object can represent an area in memory as a specific format and size. So a 64 bit process _could_ have a `SBValue` with `UINT64_MAX` children available if we wanted to have a `SBValue` that represented memory as `uint8_t` objects.
> 
> So if we are going to change stuff around, I would vote to use `uint64_t` instead of `uint32_t`

Changing the type of a return value if it changes the size/layout is ABI breaking. The mangled name wouldn't change but it may silently break something later without us realizing.

**Comment by adrian-prantl** [comment]
My main point was to unify it, but I have no problem with going for `uint64_t` instead. I'll update the patch.

File: lldb/include/lldb/Core/ValueObject.h [hunk 1]
```diff
@@ -958,7 +958,7 @@ class ValueObject {
                                           int32_t synthetic_index);
 
   /// Should only be called by ValueObject::GetNumChildren().
-  virtual size_t CalculateNumChildren(uint32_t max = UINT32_MAX) = 0;
+  virtual uint64_t CalculateNumChildren(uint64_t max = UINT32_MAX) = 0;
```
Comment (clayborg) [inline lldb/include/lldb/Core/ValueObject.h:961]: As long as we are changing things, do we want to pass in a `std::optional<uint64_t> max`? If not, then switch this to `UINT64_MAX`.

**Review by clayborg — COMMENTED** [review]
Looking good. One question: do we want to switch to using a `std::optional<uint64_t>` instead of using a `uint64_t` with a default value of UINT32_MAX? We should either use the optional or switch everything except for the public API over to use `UINT64_MAX`

**Comment by adrian-prantl** [comment]
> Looking good. One question: do we want to switch to using a `std::optional<uint64_t>` instead of using a `uint64_t` with a default value of UINT32_MAX? We should either use the optional or switch everything except for the public API over to use `UINT64_MAX`

The entire point of this patch series is to prepare for a final patch that will change the return value to `llvm::Expected<uint64_t>` :-)

File: lldb/include/lldb/Core/ValueObject.h (hunk 1 above)
Comment (adrian-prantl) [reply lldb/include/lldb/Core/ValueObject.h:961]: Oh that's separately a good change to make apart from the return value I mentioned below.

**Comment by clayborg** [comment]
> > Looking good. One question: do we want to switch to using a `std::optional<uint64_t>` instead of using a `uint64_t` with a default value of UINT32_MAX? We should either use the optional or switch everything except for the public API over to use `UINT64_MAX`
> 
> The entire point of this patch series is to prepare for a final patch that will change the return value to `llvm::Expected<uint64_t>` :-)

I was suggesting this for the input parameters only. But we can do this later.

**Review by clayborg — CHANGES_REQUESTED** [review]
So the only thing left to here is to stop using `UINT32_MAX` and switch to using `UINT64_MAX` on internal APIs.

**Comment by adrian-prantl** [comment]
Folks I'm talking this all back. Changing the return type from a uint32_t to a uint64_t is an all-or-nothing affair, because the UINT32_MAX sentinel value is permeating a bunch of other APIs that I have not intention of changing right now. For example, `GetIndexOfChildWithName()`.
I'd like to propose to just unify to `int32_t` here, which is no regression (as evidenced by the SBAPI) and then as I start threading through error handling, we can separately talk about replacing the sentinel values with something else. Then switching to 64 bits is again relatively simple.

**Comment by adrian-prantl** [comment]
So I reverted back to the state where I convert everything to int32_t. This way I don't have to touch the UINT32_MAX sentinels in this patch, which should be done separately and very carefully.

**Comment by adrian-prantl** [comment]
> I would almost vote to change everything to uint64_t except for the public API since we can't change the API without breaking. Though I winder if we can actually change this one:

@bulbazord I discovered that the idea that these are 32-bit values is deeply baked into the code by use of UINT32_MAX sentinel values in multiple places. We can of course change that too at some point, but we're really not giving anything up with this patch that isn't already a limitation.

**Review by bulbazord — APPROVED** [review]
I'm a little disappointed but I think this is the right step to take. Thanks!

**Comment by adrian-prantl** [comment]
> So the only thing left to here is to stop using UINT32_MAX and switch to using UINT64_MAX on internal APIs.

If the goal is to migrate to 64-bit everywhere, yes. But we'd need to do this on a bunch more APIs. If you feel strongly we need to do this I can do it, but the patch will get much larger this way.

**Review by clayborg — APPROVED** [review]
As long as the changes are planned for quick follow up patches I am good not changing it up front. And we can probably just switch to using `std::optional<uint64_t>` instead of switching things to use UINT32_MAX anyway with a follow up patch which is even better.

**Review by JDevlieghere — APPROVED** [review]
🚢 it!

**Comment by adrian-prantl** [comment]
Merged in https://github.com/llvm/llvm-project/commit/e710523e408ce64c15fddf9f7dbe1248795c20d7

**Pull request closed without merging**